ctpolicy/ctpolicy.go

```diff
@@ -0,0 +1,107 @@
+package ctpolicy
+
+import (
+	"context"
+	"errors"
+	// glog "log"
+	"sync"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+)
+
+// CTPolicy is used to hold information about SCTs required from various
+// groupings
+type CTPolicy struct {
+	pub    core.Publisher
+	groups [][]cmd.LogDescription
+}
+
+func New(pub core.Publisher, groups [][]cmd.LogDescription) *CTPolicy {
+	return &CTPolicy{
+		pub:    pub,
+		groups: groups,
+	}
+}
+
+// race submits an SCT to each log in a group and waits for the first response back,
+// once it has the first SCT it cancels all of the other submissions and returns.
+// It allows up to len(group)-1 of the submissions to fail as we only care about
+// getting a single SCT.
+func (ctp *CTPolicy) race(ctx context.Context, cert []byte, group []cmd.LogDescription) ([]byte, error) {
+	scts := make(chan []byte, 1)
+	errs := make(chan error, len(group))
+	subCtx, cancel := context.WithCancel(ctx)
+	for _, l := range group {
+		go func() {
+			sct, err := ctp.pub.SubmitToSingleCTWithResult(subCtx, l.URI, l.Key, cert)
+			if err != nil {
+				errs <- err
+				return
+			}
+			select {
+			case scts <- sct:
+			case <-subCtx.Done():
```

Comment:
It's a little weird to check for `.Done()` here, because there's no action to skip. AFAICT the idea is that all enqueues to `scts` after the first will block, so this provides a way out of the blocking, right? I think we actually don't need to check `subCtx.Done()` here, because the goal is to tell the deeper layers (like gRPC and HTTP) that we want to cancel.

Perhaps a more straightforward way to implement this function would be: Define a small struct containing `{sct, err}`, and make the goroutine very small: it calls `SubmitToSingleCTWithResult` and enqueues one of these structs onto a chan of size == `len(group)`. Then run the `for` loop for exactly `len(group)` iterations, and return the first non-error result. If you fall off the end of the for loop, return the last error.

Reply:
Discussed OOB: while gRPC can cancel the RPC, without checking doing a parallel check of `.Done()` the goroutine we spawn here will live on forever until the write on `scts` is unblocked, which will never happen.